### ISSUE TITLE
Issues with GatewayAPI In GKE

### ISSUE THREAD

**Issue by itsthatriver** [cilium/cilium#36028]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.3 and lower than v1.17.0

### What happened?

I created a gateway, httproute, service, and backend deployment in GKE with almost the same configuration as I have working in EKS. When all the resources were launched in GKE I found that I was unable to connect to the service through the gateway IP, with the connection hanging for a long time before failing on an ambiguous TLS error. I expected the service to be reachable at this IP/port.
```
❯ curl -vvI https://REDACTED.api-dev.dedicated.ascend.io/openapi.json
*   Trying 34.170.150.122:443...
* Connected to REDACTED.api-dev.dedicated.ascend.io (34.170.150.122) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* Recv failure: Operation timed out
* LibreSSL/3.3.6: error:02FFF03C:system library:func(4095):Operation timed out
* Closing connection
curl: (35) Recv failure: Operation timed out
```

I did a bunch of research on the exact failure details (the logs I posted below were the only failure indication I found) and they seemed to indicate that this was likely a bad cert chain, but I was unable to confirm that (all 3 parts of the cert chain were present, they appeared to validate, there were no errors in any of the gateway or route resource events...)

### How can we reproduce the issue?

1. Launch a GKE cluster (legacy datapath)
2. Install the gateway v1.1 CRDs with experimental channel for HTTPRoute and TLSRoute
3. Install Cilium on the cluster with the following values (note this is with tunneling rather than native routing as the gke get-started suggests). The values are as follows:
```
  values:
    agentNotReadyTaintKey: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-agent-not-ready
    aksbyocni:
      enabled: false
    authentication:
      mutual:
        spire:
          enabled: true
          install:
            enabled: true
            existingNamespace: true
            namespace: kube-system
    bpf:
      hostLegacyRouting: true
      masquerade: true
    cni:
      binPath: /home/kubernetes/bin
    debug:
      enabled: true
      verbose: flow envoy datapath policy
    devices: eth+
    encryption:
      enabled: true
      nodeEncryption: true
      type: wireguard
    envoy:
      enabled: true
    gatewayAPI:
      enableAlpn: true
      enableAppProtocol: true
      enabled: true
      secretsNamespace:
        create: false
        name: kube-system
    hubble:
      listenAddress: :4244
      metrics:
        enabled:
        - dns
        - drop
        - tcp
        - flow
        - icmp
        - http
      relay:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: pool
                  operator: In
                  values:
                  - default
                  - control
          podAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  k8s-app: cilium
              topologyKey: kubernetes.io/hostname
        enabled: true
      tls:
        auto:
          method: cronJob
      ui:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: pool
                  operator: In
                  values:
                  - default
                  - control
        enabled: true
    ingressController:
      enabled: false
    ipam:
      mode: kubernetes
    k8sServiceHost: REDACTEDD
    k8sServicePort: 443
    kubeProxyReplacement: true
    l7Proxy: true
    loadBalancer:
      serviceTopology: true
    localRedirectPolicy: true
    nodeinit:
      enabled: true
      reconfigureKubelet: true
      removeCbrBridge: true
    operator:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: pool
                operator: In
                values:
                - default
                - control
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                io.cilium/app: operator
            topologyKey: kubernetes.io/hostname
      prometheus:
        enabled: true
    prometheus:
      enabled: true
    upgradeCompatibility: "1.9"
    wellKnownIdentities:
      enabled: true
```
4. Set up a simple gateway:
```
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  labels:
    kustomize.toolkit.fluxcd.io/name: app-gateway
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: app-gateway
  namespace: app-gateway
spec:
  gatewayClassName: cilium
  listeners:
  - allowedRoutes:
      namespaces:
        from: All
    hostname: '*.api-dev.dedicated.ascend.io'
    name: secure-api
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - group: ""
        kind: Secret
        name: star-api-dev-dedicated-ascend-io
      mode: Terminate
```
5. Provision a cert for the gateway with cert-manager and move the cert secret to be available for the gateway listener (for simplicity it can be launched in the same namespace, or if you have a cert for testing, any cert should work here)
6. Set up an HTTPRoute on the listener
```
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  labels:
    app: instance-backend
  name: instance-httproute
  namespace: default
spec:
  hostnames:
  - REDACTED.api-dev.dedicated.ascend.io
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: app-gateway
    namespace: app-gateway
    sectionName: secure-api
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: instance-backend
      port: 80
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /
    timeouts:
      request: 120s
```
7. Set up a deployment and service behind the service to accept traffic
```
apiVersion: v1
kind: Service
metadata:
  name: instance-backend
  namespace: default
  labels:
    app: instance-backend
spec:
  selector:
    app: instance-backend
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 4200
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: instance-backend
  name: instance-backend
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: instance-backend
  template:
    metadata:
      labels:
        app: instance-backend
    spec:
      containers:
        image: nginx
        name: instance-backend
        ports:
        - containerPort: 4200
          name: api
          protocol: TCP
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - ALL
          privileged: false
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          runAsUser: 1000
          seccompProfile:
            type: RuntimeDefault
```
8. Set up DNS (or use a local etc host) and curl the gateway on 443.

### Cilium Version

❯ cilium version
cilium-cli: v0.16.18 compiled with go1.23.1 on darwin/arm64
cilium image (default): v1.16.1
cilium image (stable): v1.16.3
cilium image (running): 1.16.3

### Kernel Version

Linux 6.1.100+ #1 SMP PREEMPT_DYNAMIC Sat Aug 24 16:19:44 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

❯ kubectl version
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.5-gke.1014003

### Regression

I'm not sure if this is a regression or not. Like I mentioned above, this is working fine in EKS, but we haven't tested this exact setup in GKE (although the rest of Cilium in GKE is working great)

### Relevant log output

```shell
2024-11-18 13:09:29.818 | [2024-11-18 21:09:29.695][21][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:280] [Tags: "ConnectionId":"399"] remote address:67.188.177.91:53727,TLS_error:\|33554536:system library:OPENSSL_internal:Connection reset by peer:TLS_error_end:TLS_error_end |  
-- | -- | --
  |   | 2024-11-18 13:09:29.818 | [2024-11-18 21:09:29.695][21][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:280] [Tags: "ConnectionId":"399"] remote address:67.188.177.91:53727,TLS_error:\|33554536:system library:OPENSSL_internal:Connection reset by peer:TLS_error_end |  
  |   | 2024-11-18 13:08:19.400 | [2024-11-18 21:08:19.182][21][debug][conn_handler] [external/envoy/source/common/listener_manager/active_tcp_listener.cc:160] [Tags: "ConnectionId":"399"] new connection from 67.188.177.91:53727
```

### Anything else?

@youngnick , you mentioned to CC you for Gateway API issues, so just flagging this for your attention. I do have CiliumNetworkPolicies in the relevant namespace, but everything I'm seeing from hubble and logs is indicating that there are no drops due to policy.

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for this @thejosephstevens, I'm on triage this week as well, so I'm seeing the issue anyway.

Any chance you could upload a sysdump using the instructions at https://docs.cilium.io/en/stable/operations/troubleshooting/#automatic-log-state-collection ? That will help someone investigate this once they can get to it.

The problem does sound weird, but the sysdump allows some checking without needing to build a full repro environment first.

**Comment by thejosephstevens** [comment]
Ah, I thought I had uploaded that, but it looks like the upload failed (google seems to indicate that it was likely above the max file size limit for Github Issues upload). The sysdump is available as a tar.gz [here](https://drive.google.com/file/d/1GymlpmAEdF6_mWa-us6PPmaBHrqABQB6/view?usp=sharing).

Note, I was able to confirm that 1.16.4 does not fix the issues I'm experiencing, and I also was able to confirm that this issue is not present in AKS w/ BYOCNI (in addition to a very similar setup having no issues in EKS). All of these installs are using VXLAN encapsulation datapath. The primary differences in GKE are:
ipam_mode == kubernetes
node_init == enabled
node_init_reconfigure_kubelet == true
node_init_remove_cbr_bridge == true
cni_bin_path == /home/kubernetes/bin
bpf.host_legacy_routing == true

I'm going to test native routing in GKE to see if that addresses the issue, but past that I'm drawing a blank on what could be going on here. I did have another issue in GKE with 1.16 [here](https://github.com/cilium/cilium/issues/35977), which was addressed by using the host legacy routing, but in case that's possibly relevant I figured I'd flag here.

**Comment by thejosephstevens** [comment]
I was able to confirm today that TLSRoute also did not work. My thought was that if the issue is actually related to SSL resolution in the gateway, maybe using passthrough to the end service could avoid the problem, but that did not pan out. The symptoms of the failure were identical to with an HTTPRoute. I have not yet tested with NativeRouting.

**Comment by youngnick** [comment]
We do run CI in GKE, and I've tested Gateway API stuff there myself, so I feel confident that Gateway API stuff _can_ work in GKE, there's something here that's stopping it. I'll put this into the Service Mesh team queue and we'll work our way towards it.

**Comment by thejosephstevens** [comment]
Any chance you could point me to a working configuration so I can try to find the diff between what is and isn't working? I just looked through the conformance tests and IIUC:
* [GKE conformance tests](https://github.com/cilium/cilium/blob/main/.github/actions/gke/test-config-helm.yaml) do all use tunneling (so native routing is unlikely to be related), but none of them use wireguard.
* The Gateway API conformance tests [appear to only run on `kind`](https://github.com/cilium/cilium/blob/main/.github/workflows/conformance-gateway-api.yaml#L178-L185), so I don't see a config combination that I can mimic to validate.

I tried updating my cluster to use ipsec instead of wireguard (this required disabling host encryption and enabling dnsproxy transparent mode), but still got the same result for curl attempts. My GKE clusters appeared to otherwise be effectively identical to the test clusters launched in conformance testing, so it seems most likely that the difference is in Cilium config.

If you can get me a known good combination I can at least help isolate what change breaks the configuration. If that's not something that you have handy though, I have been able to unblock myself on this end so I can just await investigation.

**Comment by truongnht** [comment]
@thejosephstevens is the gateway cert (secret) created correctly? I used to have similar issue in EKS, turned out that the tls cert needs to be configured. In our case, we leverage cert-manager to issue a self-signed one

**Comment by thejosephstevens** [comment]
Yeah, I independently pulled the secret and verified the cert, and we're using a secret that is generated by cert-manager and delivered to the cluster by External Secret Operator. I specifically checked that earlier when debugging before submitting the issue, and when I tested the exact same setup in AKS it ran into no issues (different domain, same cert issuer). Our setup in EKS is almost completely identical as well, and also has no issues. When I tested TLS Passthrough earlier I expected to be able to hit the service itself, but traffic was still not getting past the GatewayAPI, even when it wasn't doing any TLS termination (should have just been doing SNI inspection and raw passthrough to the end service).

Separately, I would have expected some kind of error on the HTTPS listener if the cert were invalid, but I didn't see one.

**Comment by truongnht** [comment]
with EKS I got gateway api working with TLSRoute (passthrough), grpcRoute and httpRoute, haven't tried mtls yet

**Comment by thejosephstevens** [comment]
Any updates here? Anything I can try testing or information I can provide to be helpful?

**Comment by thejosephstevens** [comment]
Unmark stale, this is still a real issue

**Comment by Pionerd** [comment]
@thejosephstevens ever figured this out? Same exact issue here, so not stale imo, no permission to reopen.

**Comment by itsthatriver** [comment]
I think I was able to get it working in one instance but IIRC there were caveats on the compatibility so we never got it fully rolled out. I unfortunately no longer have access to that codebase (just changed jobs), but @geoman3 may be able to help.

**Comment by youngnick** [comment]
@Pionerd reopened, please feel free to upload a sysdump, as I mentioned earlier in the thread.

**Comment by Pionerd** [comment]
@youngnick Here you go: [cilium-sysdump-20251209-101124.zip](https://github.com/user-attachments/files/24050974/cilium-sysdump-20251209-101124.zip)

**Comment by youngnick** [comment]
@Pionerd, this doesn't really look like what @thejosephstevens described, as you're not using any TLS.

Looking at the sysdump, it looks like your hostname `argocd.internal.capetown.pionerds.nl` resolves to an RFC1918 address, but your Gateway is showing the address in its `status.addresses` as a publicly-visible, different one. Unless you're doing something with split horizon DNS, that seems like the place to start.

**Comment by Pionerd** [comment]
@youngnick as I was seeing this behavior both with TLS and without it, I assumed it would simplify things to run the sysdump without it. When using a TLS-only Gateway (like @thejosephstevens) I get the exact same errors as he posted above.

Regarding the IP: the correct public IP is present in my `/etc/hosts` file during testing, so I don't think that's related.

**Comment by youngnick** [comment]
Okay, thanks for that.

From what I can see from the Gateway API objects, and the Services, everything seems to be set up fine as far as Cilium is concerned. The Envoy config shows that the listener is present and forwarding traffic to the correct set of endpoints.

Are you also using mutual auth? I think I see some things indicating that in both of your reports. I would try without mutual auth enabled - that is very much a beta feature that has not seen the stability gains that I would have hoped for.

**Comment by Pionerd** [comment]
AFAIK we do not use mutual auth. Our Helm values below, Helm chart v1.18.4 has `authentication.mutual.spire.enabled=false` as default value.

```
agentNotReadyTaintKey: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-agent-not-ready
bandwidthManager:
  bbr: false
  enabled: true
bpf:
  hostLegacyRouting: true
  masquerade: true
cni:
  binPath: /home/kubernetes/bin
dashboards:
  annotations:
    grafana_folder: Pionative
  enabled: true
enableIPv4BIGTCP: false
encryption:
  enabled: false
  nodeEncryption: false
  type: wireguard
gatewayAPI:
  enableAlpn: false
  enabled: true
  externalTrafficPolicy: Local
  gatewayClass:
    create: "true"
  # This was a test, same behavior with trusted hops = default 0
  xffNumTrustedHops: 1
gke:
  enabled: true
hubble:
  enabled: true
  export:
    static:
      allowList:
        - '{"verdict":["DROPPED","ERROR","AUDIT"]}'
      enabled: true
      filePath: stdout
  metrics:
    dashboards:
      annotations:
        grafana_folder: Pionative
      enabled: true
    enableOpenMetrics: true
    enabled:
      - dns
      - drop
      - tcp
      - flow
      - port-distribution
      - icmp
      - kafka:labelsContext=source_namespace,source_workload,destination_namespace,destination_workload,traffic_direction;sourceContext=workload-name|reserved-identity;destinationContext=workload-name|reserved-identity
      - httpV2:exemplars=true;labelsContext=source_ip,source_namespace,source_workload,destination_ip,destination_namespace,destination_workload,traffic_direction;sourceContext=workload-name|reserved-identity;destinationContext=workload-name|reserved-identity
    serviceMonitor:
      enabled: true
  relay:
    enabled: true
    prometheus:
      enabled: true
      serviceMonitor:
        enabled: true
    resources:
      limits:
        memory: 256Mi
      requests:
        cpu: 10m
        memory: 32Mi
  ui:
    backend:
      resources:
        limits:
          memory: 1024M
        requests:
          cpu: 100m
          memory: 64Mi
    enabled: true
    frontend:
      resources:
        limits:
          memory: 1024M
        requests:
          cpu: 100m
          memory: 64Mi
    ingress:
      annotations:
        cert-manager.io/cluster-issuer: letsencrypt-prod-internal
        cert-manager.io/revision-history-limit: "1"
        nginx.ingress.kubernetes.io/auth-signin: https://oauth2.internal.capetown.pionerds.nl/oauth2/start
        nginx.ingress.kubernetes.io/auth-url: http://oauth2-proxy.oauth2-proxy.svc.cluster.local/oauth2/auth
      className: nginx-internal
      enabled: true
      hosts:
        - hubble.internal.capetown.pionerds.nl
      tls:
        - hosts:
            - hubble.internal.capetown.pionerds.nl
          secretName: hubble-tls
ipam:
  mode: kubernetes
ipv4NativeRoutingCIDR: 10.120.64.0/18
k8sServiceHost: 10.120.0.2
k8sServicePort: 443
kubeProxyReplacement: true
localRedirectPolicy: false
nodeinit:
  enabled: true
  reconfigureKubelet: true
  removeCbrBridge: true
  resources:
    requests:
      cpu: 10m
      memory: 10Mi
  restartPods: true
  startup:
    preScript: rm -f /etc/kubernetes/manifests/kube-proxy.manifest
operator:
  dashboards:
    annotations:
      grafana_folder: Pionative
    enabled: true
  podDisruptionBudget:
    enabled: true
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
  resources:
    limits:
      memory: 1Gi
    requests:
      cpu: 100m
      memory: 128Mi
policyAuditMode: false
prometheus:
  enabled: true
  serviceMonitor:
    enabled: true
resources:
  limits:
    memory: 4Gi
  requests:
    cpu: 100m
    memory: 512Mi
socketLB:
  hostNamespaceOnly: true
  terminatePodConnections: true
```
